test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -164,21 +177,48 @@ func (r *DockerMachinePoolReconciler) SetupWithManager(ctx context.Context, mgr
 			builder.WithPredicates(
 				predicates.ClusterUnpausedAndInfrastructureReady(ctrl.LoggerFrom(ctx)),
 			),
-		).Complete(r)
+		).Build(r)
 	if err != nil {
 		return errors.Wrap(err, "failed setting up with a controller manager")
 	}
+
+	r.controller = c
+	r.recorder = mgr.GetEventRecorderFor("dockermachinepool-controller")
+	r.externalTracker = external.ObjectTracker{
+		Controller: c,
+		Cache:      mgr.GetCache(),
+	}
+
 	return nil
 }
 
-func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) error {
-	pool, err := docker.NewNodePool(ctx, r.Client, cluster, machinePool, dockerMachinePool)
+func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, dockerMachinePool *infraexpv1.DockerMachinePool) error {
+	_ = ctrl.LoggerFrom(ctx)
+
+	labelFilters := map[string]string{dockerMachinePoolLabel: dockerMachinePool.Name}
+	// List Docker containers, i.e. external machines in the cluster.
+	externalMachines, err := docker.ListMachinesByCluster(ctx, cluster, labelFilters)
 	if err != nil {
-		return errors.Wrap(err, "failed to build new node pool")
+		return errors.Wrapf(err, "failed to list all machines in the cluster with label \"%s:%s\"", dockerMachinePoolLabel, dockerMachinePool.Name)
 	}
 
-	if err := pool.Delete(ctx); err != nil {
-		return errors.Wrap(err, "failed to delete all machines in the node pool")
+	// Iterate over the Docker containers, i.e. external machines, directly instead of DockerMachines to ensure none are left behind.
+	// Providers should similarly ensure that all infrastructure instances are deleted even if the InfraMachine has not been created yet.
+	for _, externalMachine := range externalMachines {
+		key := client.ObjectKey{
+			Namespace: dockerMachinePool.Namespace,
+			Name:      externalMachine.Name(),
+		}
+		dockerMachine := &infrav1.DockerMachine{}
+		if err := r.Client.Get(ctx, key, dockerMachine); err != nil {
+			if err := externalMachine.Delete(ctx); err != nil {
+				return errors.Wrapf(err, "failed to delete machine %s", externalMachine.Name())
+			}
+		}
+
+		if err := r.deleteMachinePoolMachine(ctx, *dockerMachine); err != nil {
```

Comment:
Why do we need to delete machine pool machine? When machine pools are deleted, the capi machines in it will be deleted, and since infra machines are tied to capi machines, those will also be deleted as well right?